File: ai_client_gateway/src/routes/oauth.ts
/**
 * OAuth2 Authorization Server routes for Claude Code MCP integration.
 *
 * Claude Code discovers the gateway acts as an OAuth2 AS via
 * GET /.well-known/oauth-authorization-server, then drives a
 * standard PKCE authorization_code flow. The gateway acts as a
 * thin shell that redirects to Keycloak and stores the resulting
 * ID token in a local session. The rest of the ID-JAG chain
 * (ID token → ID-JAG → Athenz AT) runs unchanged inside the proxy.
 *
 * Required env vars:
 *   KEYCLOAK_URL      e.g. http://localhost:34443
 *   KEYCLOAK_REALM    e.g. master
 *   KEYCLOAK_CLIENT_ID  e.g. ai.open-webui
 *   PUBLIC_BASE_URL   e.g. http://localhost:44443  (must be reachable by browser)
 */

import { Router, Request, Response } from "express";
import crypto from "crypto";
import { URLSearchParams } from "url";
import { PUBLIC_BASE_URL } from "../config/env.js";
import { createSession, getSession } from "../utils/sessionStore.js";

const router = Router();

const KEYCLOAK_URL = process.env.KEYCLOAK_URL ?? "http://localhost:34443";
const KEYCLOAK_PUBLIC_URL = process.env.KEYCLOAK_PUBLIC_URL ?? KEYCLOAK_URL;
const KEYCLOAK_REALM = process.env.KEYCLOAK_REALM ?? "master";
const CLIENT_ID = process.env.KEYCLOAK_CLIENT_ID ?? "claude-code";
const CLIENT_SECRET = process.env.KEYCLOAK_CLIENT_SECRET;

// code_verifier is only needed on the gateway side for state validation;
// Keycloak uses its own PKCE challenge sent from here.
const pendingStates = new Map<string, { codeVerifier: string; redirectUri: string }>();

// Server-side token exchange uses the in-cluster address.
function keycloakBase(): string {
  return `${KEYCLOAK_URL}/realms/${KEYCLOAK_REALM}/protocol/openid-connect`;
}

// Browser-facing redirects must use the publicly reachable address.
function keycloakPublicBase(): string {
  return `${KEYCLOAK_PUBLIC_URL}/realms/${KEYCLOAK_REALM}/protocol/openid-connect`;
}

// RFC 7636 helpers
function base64url(buf: Buffer): string {
  return buf.toString("base64").replace(/\+/g, "-").replace(/\//g, "_").replace(/=/g, "");
}
function generateVerifier(): string {
  return base64url(crypto.randomBytes(32));
}
function deriveChallenge(verifier: string): string {
  return base64url(crypto.createHash("sha256").update(verifier).digest());
}

// ── metadata ──────────────────────────────────────────────────────────────────

router.get("/.well-known/oauth-authorization-server", (_req: Request, res: Response) => {
  res.json({
    issuer: PUBLIC_BASE_URL,
    authorization_endpoint: `${PUBLIC_BASE_URL}/oauth/authorize`,
    token_endpoint: `${PUBLIC_BASE_URL}/oauth/token`,
    registration_endpoint: `${PUBLIC_BASE_URL}/oauth/register`,
    response_types_supported: ["code"],
    grant_types_supported: ["authorization_code"],
    code_challenge_methods_supported: ["S256"],
  });
});

// ── dynamic client registration (RFC 7591) ────────────────────────────────────
// Claude Code requires this endpoint to exist. We accept any registration and
// return a stable client_id — the actual Keycloak client is always CLIENT_ID.

router.post("/oauth/register", (req: Request, res: Response) => {
  const body = req.body as Record<string, any> ?? {};
  const issuedClientId = body.client_id ?? `mcp-client-${crypto.randomUUID()}`;

  res.status(201).json({
    client_id: issuedClientId,
    client_id_issued_at: Math.floor(Date.now() / 1000),
    redirect_uris: body.redirect_uris ?? [],
    grant_types: ["authorization_code"],
    response_types: ["code"],
    token_endpoint_auth_method: "none",
  });
});

// ── authorize ─────────────────────────────────────────────────────────────────

router.get("/oauth/authorize", (req: Request, res: Response) => {
  const { redirect_uri, state, code_challenge } = req.query as Record<string, string>;

  if (!redirect_uri || !state) {
    res.status(400).send("Missing redirect_uri or state");
    return;
  }

  const codeVerifier = generateVerifier();
  const challenge = deriveChallenge(codeVerifier);

  pendingStates.set(state, { codeVerifier, redirectUri: redirect_uri });

  const params = new URLSearchParams({
    client_id: CLIENT_ID,
    response_type: "code",
    redirect_uri: `${PUBLIC_BASE_URL}/oauth/callback`,
    scope: "openid email profile",
    state,
    code_challenge: challenge,
    code_challenge_method: "S256",
  });

  res.redirect(`${keycloakPublicBase()}/auth?${params}`);
});

// ── callback ──────────────────────────────────────────────────────────────────

router.get("/oauth/callback", async (req: Request, res: Response) => {
  const { code, state, error: kErr } = req.query as Record<string, string>;

  if (kErr) {
    res.status(400).send(`Keycloak error: ${kErr}`);
    return;
  }

  const pending = state ? pendingStates.get(state) : null;
  if (!pending || !code) {
    res.status(400).send("Invalid state or missing code");
    return;
  }
  pendingStates.delete(state);

  try {
    // Exchange the Keycloak code for tokens
    const body = new URLSearchParams({
      grant_type: "authorization_code",
      client_id: CLIENT_ID,
      ...(CLIENT_SECRET ? { client_secret: CLIENT_SECRET } : {}),
      code,
      redirect_uri: `${PUBLIC_BASE_URL}/oauth/callback`,
      code_verifier: pending.codeVerifier,
    });

    const tokenRes = await fetch(`${keycloakBase()}/token`, {
      method: "POST",
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
      body: body.toString(),
    });

    if (!tokenRes.ok) {
      const text = await tokenRes.text();
      res.status(502).send(`Keycloak token exchange failed: ${tokenRes.status} ${text}`);
      return;
    }

    const tokenData: any = await tokenRes.json();
    const idToken: string = tokenData.id_token;

    if (!idToken) {
      res.status(502).send("Keycloak did not return an id_token");
      return;
    }

    // Parse expiry from id_token payload
    let exp = Math.floor(Date.now() / 1000) + 3600;
    try {
      const payload = JSON.parse(Buffer.from(idToken.split(".")[1], "base64").toString());
      if (payload.exp) exp = payload.exp;
    } catch { /* keep default */ }

    const sessionToken = createSession(idToken, exp);

    // Redirect back to Claude Code's redirect_uri with our session token as the code
    const callbackParams = new URLSearchParams({ code: sessionToken, state });
    res.redirect(`${pending.redirectUri}?${callbackParams}`);
  } catch (err: any) {
    res.status(500).send(`Callback error: ${err.message}`);
  }
});

// ── token ─────────────────────────────────────────────────────────────────────

router.post("/oauth/token", (req: Request, res: Response) => {
  // Claude Code posts the code (which is actually our session token) here
  const body = req.body as Record<string, string>;
  const code = body.code ?? (typeof req.body === "string"
    ? new URLSearchParams(req.body).get("code")
    : null);

  if (!code) {
    res.status(400).json({ error: "invalid_request", error_description: "Missing code" });
    return;
  }

  const session = getSession(code);
  if (!session) {
    res.status(400).json({ error: "invalid_grant", error_description: "Code not found or expired" });
    return;
  }

  const expiresIn = Math.max(0, session.exp - Math.floor(Date.now() / 1000));

  res.json({
    access_token: code,      // re-use the session token as the bearer
    token_type: "Bearer",
    expires_in: expiresIn,
    scope: "openid email profile",
  });
});

export default router;

File: showcases/mcp_x_idjag/components/mcp-bff-gateway/oauth.go
// OAuth2 Authorization Server endpoints for the Pattern 3b BFF.
//
// The real Client (Claude Code, or any generic MCP client) discovers this
// gateway as the AS via GET /.well-known/oauth-authorization-server, then
// drives a standard PKCE authorization_code flow against it. This gateway is
// itself Keycloak's OIDC RP: it redirects the Client's browser to Keycloak,
// completes the login on the Client's behalf, and stores the resulting
// id_token in a server-side session. The Client is only ever handed the
// opaque session id, never the id_token and never a real Athenz Access
// Token - satisfying Pattern 3b's "remote id_token acquisition, AT never
// returned to the Client" design.
package main

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"
)

// pendingAuth holds the real Client's own PKCE state (its redirect_uri) just
// long enough to complete the round trip through Keycloak.
type pendingAuth struct {
	codeVerifier string
	redirectURI  string
}

type oauthServer struct {
	publicBaseURL     string
	keycloakURL       string // server-to-server calls (in-cluster address)
	keycloakPublicURL string // where the human's browser is redirected
	keycloakRealm     string
	keycloakClientID  string // this gateway's own Keycloak client, not the Client's

	sessions *sessionStore

	mu      sync.Mutex
	pending map[string]pendingAuth
}

func newOAuthServer(publicBaseURL, keycloakURL, keycloakPublicURL, keycloakRealm, keycloakClientID string, sessions *sessionStore) *oauthServer {
	return &oauthServer{
		publicBaseURL:     publicBaseURL,
		keycloakURL:       keycloakURL,
		keycloakPublicURL: keycloakPublicURL,
		keycloakRealm:     keycloakRealm,
		keycloakClientID:  keycloakClientID,
		sessions:          sessions,
		pending:           make(map[string]pendingAuth),
	}
}

func (o *oauthServer) keycloakBase() string {
	return fmt.Sprintf("%s/realms/%s/protocol/openid-connect", o.keycloakURL, o.keycloakRealm)
}

func (o *oauthServer) keycloakPublicBase() string {
	return fmt.Sprintf("%s/realms/%s/protocol/openid-connect", o.keycloakPublicURL, o.keycloakRealm)
}

func (o *oauthServer) registerRoutes(mux *http.ServeMux) {
	mux.HandleFunc("GET /.well-known/oauth-authorization-server", o.handleMetadata)
	mux.HandleFunc("POST /oauth/register", o.handleRegister)
	mux.HandleFunc("GET /oauth/authorize", o.handleAuthorize)
	mux.HandleFunc("GET /oauth/callback", o.handleCallback)
	mux.HandleFunc("POST /oauth/token", o.handleToken)
}

func (o *oauthServer) handleMetadata(w http.ResponseWriter, _ *http.Request) {
	writeJSON(w, http.StatusOK, map[string]any{
		"issuer":                           o.publicBaseURL,
		"authorization_endpoint":           o.publicBaseURL + "/oauth/authorize",
		"token_endpoint":                   o.publicBaseURL + "/oauth/token",
		"registration_endpoint":            o.publicBaseURL + "/oauth/register",
		"response_types_supported":         []string{"code"},
		"grant_types_supported":            []string{"authorization_code"},
		"code_challenge_methods_supported": []string{"S256"},
	})
}

// handleRegister implements RFC 7591 dynamic client registration. Generic MCP
// clients (Claude Code included) expect this endpoint to exist; we accept any
// registration and return a stable client_id.
func (o *oauthServer) handleRegister(w http.ResponseWriter, r *http.Request) {
	var body struct {
		ClientID     string   `json:"client_id"`
		RedirectURIs []string `json:"redirect_uris"`
	}
	_ = json.NewDecoder(r.Body).Decode(&body)

	clientID := body.ClientID
	if clientID == "" {
		clientID = "mcp-client-" + randomID()
	}
	redirectURIs := body.RedirectURIs
	if redirectURIs == nil {
		redirectURIs = []string{}
	}

	writeJSON(w, http.StatusCreated, map[string]any{
		"client_id":                  clientID,
		"client_id_issued_at":        time.Now().Unix(),
		"redirect_uris":              redirectURIs,
		"grant_types":                []string{"authorization_code"},
		"response_types":             []string{"code"},
		"token_endpoint_auth_method": "none",
	})
}

func (o *oauthServer) handleAuthorize(w http.ResponseWriter, r *http.Request) {
	q := r.URL.Query()
	redirectURI := q.Get("redirect_uri")
	state := q.Get("state")
	if redirectURI == "" || state == "" {
		http.Error(w, "Missing redirect_uri or state", http.StatusBadRequest)
		return
	}

	verifier := generateVerifier()
	challenge := deriveChallenge(verifier)

	o.mu.Lock()
	o.pending[state] = pendingAuth{codeVerifier: verifier, redirectURI: redirectURI}
	o.mu.Unlock()

	params := url.Values{}
	params.Set("client_id", o.keycloakClientID)
	params.Set("response_type", "code")
	params.Set("redirect_uri", o.publicBaseURL+"/oauth/callback")
	params.Set("scope", "openid email profile")
	params.Set("state", state)
	params.Set("code_challenge", challenge)
	params.Set("code_challenge_method", "S256")

	http.Redirect(w, r, o.keycloakPublicBase()+"/auth?"+params.Encode(), http.StatusFound)
}

func (o *oauthServer) handleCallback(w http.ResponseWriter, r *http.Request) {
	q := r.URL.Query()
	if kErr := q.Get("error"); kErr != "" {
		http.Error(w, "Keycloak error: "+kErr, http.StatusBadRequest)
		return
	}

	code := q.Get("code")
	state := q.Get("state")

	o.mu.Lock()
	pending, ok := o.pending[state]
	if ok {
		delete(o.pending, state)
	}
	o.mu.Unlock()

	if !ok || code == "" {
		http.Error(w, "Invalid state or missing code", http.StatusBadRequest)
		return
	}

	form := url.Values{}
	form.Set("grant_type", "authorization_code")
	form.Set("client_id", o.keycloakClientID)
	form.Set("code", code)
	form.Set("redirect_uri", o.publicBaseURL+"/oauth/callback")
	form.Set("code_verifier", pending.codeVerifier)

	tokenReq, err := http.NewRequest(http.MethodPost, o.keycloakBase()+"/token", strings.NewReader(form.Encode()))
	if err != nil {
		http.Error(w, "Failed to build Keycloak token request: "+err.Error(), http.StatusInternalServerError)
		return
	}
	tokenReq.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	// Keycloak derives each issued token's "iss" claim from the Host header of
	// this request, not from KEYCLOAK_URL's actual network target. Athenz ZTS
	// only recognizes the browser-facing KEYCLOAK_PUBLIC_URL as a trusted
	// issuer (see bootstrap-common/03-athenz.sh's issuerUri), so this request
	// must present that Host even though it physically connects to the
	// in-cluster KEYCLOAK_URL - otherwise ZTS cannot map the token's identity
	// to an Athenz principal and every role check fails.
	if publicHost := hostOf(o.keycloakPublicURL); publicHost != "" {
		tokenReq.Host = publicHost
	}

	resp, err := http.DefaultClient.Do(tokenReq)
	if err != nil {
		http.Error(w, "Keycloak token exchange failed: "+err.Error(), http.StatusBadGateway)
		return
	}
	defer resp.Body.Close()

	var tokenBody struct {
		IDToken string `json:"id_token"`
	}
	decodeErr := json.NewDecoder(resp.Body).Decode(&tokenBody)
	if resp.StatusCode >= 300 {
		http.Error(w, fmt.Sprintf("Keycloak token exchange failed: %d", resp.StatusCode), http.StatusBadGateway)
		return
	}
	if decodeErr != nil || tokenBody.IDToken == "" {
		http.Error(w, "Keycloak did not return an id_token", http.StatusBadGateway)
		return
	}

	exp := time.Now().Add(time.Hour).Unix()
	if claims, err := decodeJWTPayload(tokenBody.IDToken); err == nil {
		if e, ok := claims["exp"].(float64); ok {
			exp = int64(e)
		}
	}

	sessionID := o.sessions.create(tokenBody.IDToken, exp)

	// Redirect back to the real Client's redirect_uri with the session id as
	// the authorization code. Neither the id_token nor any Athenz token ever
	// reaches this response.
	redirectParams := url.Values{}
	redirectParams.Set("code", sessionID)
	redirectParams.Set("state", state)
	http.Redirect(w, r, pending.redirectURI+"?"+redirectParams.Encode(), http.StatusFound)
}

func (o *oauthServer) handleToken(w http.ResponseWriter, r *http.Request) {
	if err := r.ParseForm(); err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "invalid_request"})
		return
	}
	code := r.FormValue("code")
	if code == "" {
		writeJSON(w, http.StatusBadRequest, map[string]string{
			"error": "invalid_request", "error_description": "Missing code",
		})
		return
	}

	sess, ok := o.sessions.get(code)
	if !ok {
		writeJSON(w, http.StatusBadRequest, map[string]string{
			"error": "invalid_grant", "error_description": "Code not found or expired",
		})
		return
	}

	expiresIn := sess.exp - time.Now().Unix()
	if expiresIn < 0 {
		expiresIn = 0
	}

	writeJSON(w, http.StatusOK, map[string]any{
		// The bearer credential handed back is the opaque session id, never a
		// real Athenz Access Token or the underlying id_token.
		"access_token": code,
		"token_type":   "Bearer",
		"expires_in":   expiresIn,
		"scope":        "openid email profile",
	})
}

func writeJSON(w http.ResponseWriter, status int, body any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(body)
}

// RFC 7636 helpers.

func generateVerifier() string {
	b := make([]byte, 32)
	if _, err := rand.Read(b); err != nil {
		panic(err)
	}
	return base64.RawURLEncoding.EncodeToString(b)
}

func deriveChallenge(verifier string) string {
	sum := sha256.Sum256([]byte(verifier))
	return base64.RawURLEncoding.EncodeToString(sum[:])
}

func decodeJWTPayload(token string) (map[string]any, error) {
	parts := strings.Split(token, ".")
	if len(parts) != 3 {
		return nil, fmt.Errorf("invalid JWT")
	}
	payload, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		return nil, err
	}
	var claims map[string]any
	if err := json.Unmarshal(payload, &claims); err != nil {
		return nil, err
	}
	return claims, nil
}

func hostOf(rawURL string) string {
	u, err := url.Parse(rawURL)
	if err != nil {
		return ""
	}
	return u.Host
}
